File: codes/Shuofei/week 3/111_Minimum Depth of Binary Tree.java
// method 1: dfs
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        if ((root.left == null) && (root.right == null)) return 1;
        if (root.left == null) return minDepth(root.right) + 1;
        if (root.right == null) return minDepth(root.left) + 1;
        return Math.min(minDepth(root.left), minDepth(root.right)) + 1;
    }
}

// method 2: bfs, more efficient 
class Solution {
    public int minDepth(TreeNode root) {
        if (root == null) return 0;
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        int res = 1;
        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                TreeNode node = q.poll();
                if ((node.left == null) && (node.right == null)) return res;
                if (node.left != null) q.offer(node.left);          
                if (node.right != null) q.offer(node.right);                     
            }
            res++;
        }
        return res;
    }
}

File: codes/Aiamjay/Week3-Day4/111. Minimum Depth of Binary Tree.py
# encoding= utf-8
import math

from codes.Aiamjay.utils.treenode import (
    TreeNode, create_tree_from_list,
    print_tree
)


class Solution:
    def minDepth(self, root: TreeNode) -> int:

        def helper(root):
            if not root.left and not root.right:
                return 1
            return 1 + min(self.minDepth(root.left) if root.left else math.inf,
                           self.minDepth(root.right) if root.right else math.inf)

        if not root:
            return 0
        return helper(root)

    def test_solution(self):
        # case 1
        print("case1")
        a = [3, 9, 20, None, None, 15, 7]
        root = create_tree_from_list(a)
        print_tree(root)
        print(self.minDepth(root))

        # case 2
        print("case2")
        a = [3, None, None]
        root = create_tree_from_list(a)
        print_tree(root)
        print(self.minDepth(root))

        # case3
        print("case3")
        a = [None]
        root = create_tree_from_list(a)
        print_tree(root)
        print(self.minDepth(root))


if __name__ == '__main__':
    s = Solution()
    s.test_solution()
    pass
